# Q05: 09/Jun/2022

1) You are watching French Open at Roland Garros. There is a 85% chance that Rafael Nadal will win the tournament. Then what are his odds?

2) Consider the data shared in earlier examples to predict leukemia remissions that patients may have(refer to dataset in GitHub - LR2.csv). The description of variables are as follows (again for your quick reference): 
* Outcome variable REMISS stands for remission: If lukemia remission occurred then 1. Else 0.
* Other predictor variables are as follows:
    * CELL: cellularity of the marrow clot section
    * SMEAR: smear differential percentage of blasts
    * INFIL: percentage of absolute marrow leukemia cell infiltrate
    * LI: percentage labeling index of the bone marrow leukemia cells
    * BLAST: absolute number of blasts in the peripheral blood
    * TEMP: the highest temperature prior to start of treatment
  
  
* What is the regression output while considering one variable "LI" (which you have all found as most significant variable in earlier question) between Logit and GLM models? Are they same?
* What is the Log Likelihood Ratio test statistic? (Hint: this is -2 * (LL-null - LL-full))

## 1

In [1]:
import numpy as np
probability_winning = 0.85
odd = probability_winning / (1- probability_winning)
odd = np.round(odd, 2)
print(f"Odds of winning for Rafael Nadal is {odd}")

Odds of winning for Rafael Nadal is 5.67


## 2

In [2]:
import pandas as pd 
import numpy as np
import statsmodels.api as sm
from statsmodels.api import Logit, add_constant
from statsmodels.stats.outliers_influence import variance_inflation_factor

df = pd.read_csv('LR2.csv')
# df.head(2)

# check for multi-collinearity before considering variables for modelling
def calculate_vif(df):
    vif = pd.DataFrame()
    vif["variables"] = df.columns
    vif["VIF"] = [variance_inflation_factor(df.values, i) for i in range(df.shape[1])]
    vif.sort_values('VIF', ascending=False, inplace=True)
    return(vif)

def fit_logistic_reg_with_intercept(X, Y):
    X = sm.add_constant(X) # adding a constant
    log_reg_model = sm.Logit(y, X).fit()
    return log_reg_model

### Logit model

In [3]:
x_var = ['LI']
y_var = ['REMISS']

X = df[x_var]
y = df[y_var]
log_model = fit_logistic_reg_with_intercept(X, y)
print(log_model.summary())

Optimization terminated successfully.
         Current function value: 0.482833
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:                 REMISS   No. Observations:                   27
Model:                          Logit   Df Residuals:                       25
Method:                           MLE   Df Model:                            1
Date:                Thu, 09 Jun 2022   Pseudo R-squ.:                  0.2414
Time:                        22:15:06   Log-Likelihood:                -13.036
converged:                       True   LL-Null:                       -17.186
Covariance Type:            nonrobust   LLR p-value:                  0.003967
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -3.7771      1.379     -2.740      0.006      -6.479      -1.075
LI             2.8973      1.

### GLM using Link

In [4]:
# Fit the model. Use GLM
glm_model = sm.GLM.from_formula("REMISS ~ LI", family=sm.families.Binomial(), data=df)
glm_model = glm_model.fit()
print(glm_model.summary())

                 Generalized Linear Model Regression Results                  
Dep. Variable:                 REMISS   No. Observations:                   27
Model:                            GLM   Df Residuals:                       25
Model Family:                Binomial   Df Model:                            1
Link Function:                  Logit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -13.036
Date:                Thu, 09 Jun 2022   Deviance:                       26.073
Time:                        22:15:06   Pearson chi2:                     23.9
No. Iterations:                     4   Pseudo R-squ. (CS):             0.2646
Covariance Type:            nonrobust                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -3.7771      1.379     -2.740      0.0

### 2a
What is the regression output while considering one variable "LI" (which you have all found as most significant variable in earlier question) between Logit and GLM models? Are they same?

Both models have similar coefficients estimates.

In [5]:
# Using model summry print output, we have LL_Null and LL_Full
# Copy paste the values and calculate the test statistics

LL_Null = -17.186
LL_Full = -13.036

-2 * (LL_Null - LL_Full)

8.3

### 2b
What is the Log Likelihood Ratio test statistic? (Hint: this is -2 * (LL-null - LL-full))

8.3